File: AzimMuradov/Term4/Block1/Task1/Fibers/Fibers/Process.cs
namespace Fibers;

public class Process
{
    private static readonly Random Rng = new();


    private readonly ProcessManager pm;

    private const int LongPauseBoundary = 10000;

    private const int ShortPauseBoundary = 100;

    private const int WorkBoundary = 1000;

    private const int IntervalsAmountBoundary = 10;
    private const int PriorityLevelsNumber = 10;

    private readonly List<int> _workIntervals = new();
    private readonly List<int> _pauseIntervals = new();

    public Process(ProcessManager pm)
    {
        this.pm = pm;

        var amount = Rng.Next(IntervalsAmountBoundary);

        for (var i = 0; i < amount; i++)
        {
            _workIntervals.Add(Rng.Next(WorkBoundary));
            _pauseIntervals.Add(Rng.Next(Rng.NextDouble() > 0.9 ? LongPauseBoundary : ShortPauseBoundary));
        }

        Priority = Rng.Next(PriorityLevelsNumber);
    }

    public void Run()
    {
        for (var i = 0; i < _workIntervals.Count; i++)
        {
            Thread.Sleep(_workIntervals[i]); // work emulation
            var pauseBeginTime = DateTime.Now;
            do
            {
                pm.Switch(isFiberFinished: false);
                Thread.Sleep(30); // reduces work on PC
            } while ((DateTime.Now - pauseBeginTime).TotalMilliseconds < _pauseIntervals[i]); // I/O emulation
        }
        pm.Switch(isFiberFinished: true);
    }

    public int Priority { get; private set; }

    public int TotalDuration => ActiveDuration + _pauseIntervals.Sum();

    public int ActiveDuration => _workIntervals.Sum();
}

File: AzimMuradov/Term4/Block1/Task1/lib/src/main/kotlin/fibers/Process.kt
package fibers

import kotlin.random.Random
import kotlin.random.nextUInt


class Process(private val pm: ProcessManager) {

    var priority: UInt = Random.nextUInt(range = 1u..PRIORITY_LEVELS_NUMBER)
        private set

    private val intervals: List<Interval> = List(size = Random.nextInt(INTERVALS_AMOUNT_BOUNDARY)) {
        Random.nextLong(WORK_BOUNDARY) to Random.nextLong(
            if (Random.nextDouble() > 0.9) LONG_PAUSE_BOUNDARY else SHORT_PAUSE_BOUNDARY
        )
    }

    val activeDuration: Long = intervals.sumOf { it.workTime }

    val totalDuration: Long = intervals.sumOf { it.totalTime }


    @Suppress("BlockingMethodInNonBlockingContext")
    fun run() {
        for ((workTime, pauseTime) in intervals) {
            Thread.sleep(workTime) // work emulation
            val pauseBeginTime = System.currentTimeMillis()
            do {
                pm.switch(isFiberFinished = false)
            } while (System.currentTimeMillis() - pauseBeginTime < pauseTime) // I/O emulation
        }
        pm.switch(isFiberFinished = true)
    }


    companion object {

        private const val PRIORITY_LEVELS_NUMBER = 10u

        private const val INTERVALS_AMOUNT_BOUNDARY = 10
        private const val SHORT_PAUSE_BOUNDARY = 100L
        private const val LONG_PAUSE_BOUNDARY = 100L
        private const val WORK_BOUNDARY = 1000L


        private val Interval.workTime: Long get() = first
        private val Interval.pauseTime: Long get() = second
        private val Interval.totalTime: Long get() = workTime + pauseTime
    }
}

private typealias Interval = Pair<Long, Long>
